File: StringToINTERCAL.js
/**
 * StringToINTERCAL.js
 *
 * Converts a string to an INTERCAL script that will output said string.
 * usage: node StringToINTERCAL.js your string here
 *
 * @author Maxamilian Demian <max@maxdemian.com>
 * @link https://www.maxodev.org
 * @link https://github.com/Maxoplata/StringToINTERCAL
 */

// class definition
class StringToINTERCAL {
	constructor() {
		this.politeCount = 0
	}

	politeLine(line) {
		if (this.politeCount === 3) {
			this.politeCount = 0;

			return `PLEASE ${line}\n`;
		}

		this.politeCount++;

		return `DO ${line}\n`;
	}

	leadingZeros(dec) {
		const count = dec.length;

		if (count < 8) {
			dec = '0'.repeat(8 - count) + dec;
		}

		return dec;
	}

	convertToINTERCAL(string) {
		// reset politeCount
		this.politeCount = 0;

		let ret = this.politeLine(`,1 <- #${string.length}`);

		let lastCharLoc = 256;
		[...string].forEach((char, i) => {
			const charLoc = parseInt(this.leadingZeros(char.charCodeAt(0).toString(2)).split('').reverse().join(''), 2)

			let movePosition = 0;

			if (charLoc < lastCharLoc) {
				movePosition = (lastCharLoc - charLoc);
			} else if (charLoc > lastCharLoc) {
				movePosition = (256 - charLoc) + lastCharLoc;
			}

			lastCharLoc -= movePosition;
			if (lastCharLoc < 1) {
				lastCharLoc = 256 + lastCharLoc;
			}

			ret += this.politeLine(`,1 SUB #${(i + 1)} <- #${movePosition}`);
		});

		ret += this.politeLine('READ OUT ,1');
		ret += this.politeLine('GIVE UP');

		return ret;
	}
}

// 1337 codez
if (process.argv.length > 2) {
	const inputString = process.argv.splice(2).join(' ');

	const myINTERCAL = new StringToINTERCAL();

	console.log(myINTERCAL.convertToINTERCAL(inputString));
}

File: StringToINTERCAL.java
/**
 * StringToINTERCAL.java
 *
 * Converts a string to an INTERCAL script that will output said string.
 * usage: javac StringToINTERCAL.java && java StringToINTERCAL your string here
 *
 * @author Maxamilian Demian <max@maxdemian.com>
 * @link https://www.maxodev.org
 * @link https://github.com/Maxoplata/StringToINTERCAL
 */

class StringToINTERCAL {
	static int politeCount = 0;

	static String politeLine(String line) {
		if (politeCount == 3) {
			politeCount = 0;

			return String.format("PLEASE %s\n", line);
		}

		politeCount++;

		return String.format("DO %s\n", line);
	}

	static String leadingZeros(String dec) {
		int count = dec.length();

		if (count < 8) {
			dec = new String(new char[8 - count]).replace("\0", "0") + dec;
		}

		return dec;
	}

	static String convertToINTERCAL(String str) {
		// reset politeCount
		politeCount = 0;

		String ret = politeLine(String.format(",1 <- #%s", str.length()));

		int lastCharLoc = 256;
		for (int i = 0; i < str.length(); i++) {
			int charLoc = Integer.parseInt(new StringBuilder(leadingZeros(Integer.toBinaryString((int) str.charAt(i)))).reverse().toString(), 2);

			int movePosition = 0;

			if (charLoc < lastCharLoc) {
				movePosition = (lastCharLoc - charLoc);
			} else if (charLoc > lastCharLoc) {
				movePosition = (256 - charLoc) + lastCharLoc;
			}

			lastCharLoc -= movePosition;
			if (lastCharLoc < 1) {
				lastCharLoc = 256 + lastCharLoc;
			}

			ret += politeLine(String.format(",1 SUB #%d <- #%d", (i + 1), movePosition));
		}

		ret += politeLine("READ OUT ,1");
		ret += politeLine("GIVE UP");

		return ret;
	}

	public static void main(String[] args) {
		if (args.length > 0) {
			String inputString = String.join(" ", args);

			System.out.println(convertToINTERCAL(inputString));
		}
	}
}
